pkg/dynamicinstrumentation/codegen/capture_depth.go

```diff
@@ -0,0 +1,151 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+//go:build linux_bpf
+
+// Package codegen is used to generate bpf program source code based on probe definitions
+package codegen
+
+import (
+	"reflect"
+
+	"github.com/DataDog/datadog-agent/pkg/dynamicinstrumentation/ditypes"
+)
+
+type captureDepthItem struct {
+	depth     int
+	parameter *ditypes.Parameter
+}
+
+func applyCaptureDepth(params []*ditypes.Parameter, targetDepth int) []*ditypes.Parameter {
+	setDoNotCapture(params, targetDepth)
+	params = pruneDoNotCaptureParams(params)
+	correctStructSizes(params)
+	return params
+}
+
+// pruneDoNotCaptureParams removes any parameters with DoNotCapture set to true from the parameter tree
+func pruneDoNotCaptureParams(params []*ditypes.Parameter) []*ditypes.Parameter {
+	if len(params) == 0 {
+		return params
+	}
+
+	result := []*ditypes.Parameter{}
+	for _, param := range params {
+		if param == nil || param.DoNotCapture {
+			continue
+		}
+
+		// Recursively prune child parameters
+		if len(param.ParameterPieces) > 0 {
+			param.ParameterPieces = pruneDoNotCaptureParams(param.ParameterPieces)
+		}
+
+		result = append(result, param)
+	}
+
+	return result
+}
+
+func setDoNotCapture(params []*ditypes.Parameter, targetDepth int) {
+	queue := []*captureDepthItem{}
+	for i := range params {
+		if params[i] == nil {
+			continue
+		}
+		queue = append(queue, &captureDepthItem{
+			depth:     1,
+			parameter: params[i],
+		})
+	}
+	for len(queue) > 0 {
+		top := queue[0]
+		queue = queue[1:]
+
+		if top.parameter == nil {
+			continue
+		}
+
+		if top.depth > targetDepth {
+			top.parameter.DoNotCapture = true
+			top.parameter.NotCaptureReason = ditypes.CaptureDepthReached
+		}
+
+		if top.parameter.Kind == uint(reflect.Struct) || top.parameter.Kind == uint(reflect.Array) {
```

Comment:
Since each of your if-statements are look at `top.parameter.Kind`, it might make sense to switch this to a switch statement